docs/design-proposals/multi-tenancy/multi-tenancy-network.md

```diff
@@ -120,10 +128,118 @@ The reasons that DNS service is per-network instead of per-space are:
 
+#### Type definition
+
+Addition of service spec is metadata.network field. By default, it is the default network of the tenant. 
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  network: my-network
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+```
+
+#### Naming Problem & Solution
+
+Services, as one of the critical network related resource types, are withing the network boundary. There is no such *shared* services of tenant or whole cluster, as both the critical virtual IP address and the associated endpoints (typically leading to pods) are all of same network.
+
+Putting service type in tenant level is out of primacy as it has following flaws:
+1. service IP address has to be the same across all networks, which implies some range of service IP needs to be predefined as shared across networks of a tenant (tenant manageability burden);  
+1. associating network-agnostic service to network-specific endpoints would introduce significant inconsistency;
+1. kubectl query for service (e.g. ```kubectl describe service```) has no way to return meaningful content of endpoints.
+
+The well-known services (default/kubernetes, kube-system/kube-dns) present some problem here. Service kubernetes has tenant/namespace(default) scope in current implementation, unable to have multiple kubernets services with each of them in one network due to name conflict. The same rational applies to kube-dns.
+
+One solution is to making its name colliding space to tenant+namespace+network, which is also breaking and inconsistent with other types (e.g. introducing network as one more parameter to call kubectl get service).
+
+Another solution is to change their name to schema that can uniquely differentiate by combining with network name; this is awkward and breaking in general. 
+
+One plausible remedy is making stringent containment of network to namespaces - resources in one namespace belongs solely to one network; one network can accommodate multiple namespace. This cannot handle the default/kubernetes service problem, as it is of default namespace.
+
+None of the two of above seems ideal. However, the latter, fiddling with the service name, if coupled with trick applicable only to kubernetes & kube-dns, seems pragmatic and practical. The idea is suffixing service name with network name (e.g. kubernetes_mynetwork), also answering query for kubernetes with that of kubernetes_mynetwork (by alias record or other form of dynamic technique), so that pods are able to look up them without using the network-suffixed names. This approach is not flexible and depending on dns ability; it works fine with limited number of well-known aliases, though. coreDNS, one of the most commonly used dns server binaries, has such dynamic aliasing ability.
```

Comment:
If I understand this correctly the suffix issue, leaks service names across tenants?

Reply:
yes as the name is well-known and fixed, objects (e.g. pods) of each tenant knows of (and actually relies on) these fixed-named service. -- name leakage, in some sense - for good reason.
However, each tenant is strongly isolated; it is impossible for one in a tenant to get the content of such service of another tenant - it can only get hold of the one of its tenant.